backstory/src/main/java/com/google/sps/story/data/NLServiceClient.java

```diff
@@ -0,0 +1,195 @@
+   * in a list linked to their relevant word type (e.g. "Mary" would
+   * be grouped into the PROPER_NOUN list and "calm" would be grouped
+   * into the adjective list).
+   *
+   * @param words the words to group by word type
+   * @return a map with lists of words grouped by their word type and with their
+   *    word type as a key
+   */
+  public Map<WordType, List<String>> groupByWordType(List<String> words) {
+    Map<WordType, List<String>> map = new HashMap<WordType, List<String>>();
+
+    List<String> singleWords = new ArrayList<String>();
+
+    // remove multi-word inputs first and store as multiword nouns
+    for (String text : words) {
+      if (text.contains(" ")) {
+        List<String> wordList = map.get(WordType.MULTIWORD_NOUN);
+
+        if (wordList == null) {
+          wordList = new ArrayList<String>();
+          map.put(WordType.MULTIWORD_NOUN, wordList);
+        }
+
+        wordList.add(text);
+
+      } else {
+        singleWords.add(text);
+      }
+    }
+
+    // TODO: Parallelize requests to Cloud Natural Language API
+    
+    for (String word : singleWords) {
+      Document doc = buildDocumentFromText(word);
+
+      AnalyzeSyntaxResponse response = client.analyzeSyntax(doc);
+
+      Token token = response.getTokens(0);
+      PartOfSpeech partOfSpeech = token.getPartOfSpeech();
+
+      WordType type = WordType.UNUSABLE;
+
+      switch (partOfSpeech.getTag()) {
+        case VERB:
+          if (isGerund(word)) {
+            type = WordType.GERUND;
+          }
+          break;
+        case NOUN:
+          if (isGerund(word)) {
+            type = WordType.GERUND;
+            break;
+          }
+
+          if (partOfSpeech.getProper() == Proper.PROPER) {
+            type = WordType.PROPER_NOUN;
+          } else {
+            type = WordType.NOUN;
+          }
+          break;
+
+        case ADJ:
+          type = WordType.ADJECTIVE;
+          break;
+      }
+
+      List<String> wordList = map.get(type);
+
+      if (wordList == null) {
+        wordList = new ArrayList<String>();
+        map.put(type, wordList);
+      }
+        
+      wordList.add(word);
+    }
+
+    return map;
+  }
+
+  /**
+   * Build and return a document from the given test.
+   *
+   * @param text the text to build a document from
+   * @return a Document that holds the current text
+   */
+  private Document buildDocumentFromText(String text) {
+    Document doc = Document.newBuilder().setContent(text).setType(Type.PLAIN_TEXT).build();
+
+    return doc;
+  }
+
+  /**
+   * Checks if a word is a gerund using heuristics (ending in "ing)
+   * and then making a call to the NL API if the word passes
+   * the heuristic test.
+   *
+   * @param word the word to check if it's a gerund
+   * @return true, if it's a gerund; false, otherwise
+   */
+  private boolean isGerund(String word) {
+    // first, check it's one word
+    if (word.contains(" ")) {
```

Comment:
Nit: consider question about whitespace mentioned above

Reply:
In reference to the whitespace question, I think I'm actually going to put those words in unusable. This class was created mainly for the purpose of parsing words sent in from Cloud Vision, which don't have whitespace in them, and so a word formatted that way, I don't think is usable in our current prompt creation suite.

Reply:
Done